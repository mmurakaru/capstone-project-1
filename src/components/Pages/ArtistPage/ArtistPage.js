import styled from 'styled-components/macro'
import PropTypes from 'prop-types'
import AudioButton from '../../Buttons/AudioButton'
import Button from '../../Buttons/Button'

ArtistPage.propTypes = {
  artists: PropTypes.array.isRequired,
  toggleCurrentSongId: PropTypes.func.isRequired,
  isSongPlaying: PropTypes.bool.isRequired,
  currentSongId: PropTypes.string.isRequired,
  saveSong: PropTypes.func.isRequired,
  savedSongsList: PropTypes.array.isRequired,
}

export default function ArtistPage({
  artists,
  toggleCurrentSongId,
  isSongPlaying,
  currentSongId,
  saveSong,
  savedSongsList,
}) {
  return (
    <ListStyled>
      {artists?.map(({ artist, songTitle, id }) => (
        <ListItemStyled key={id}>
          <Artist>{artist}</Artist>
          <Song>{songTitle}</Song>
          <AudioButtonWrapper>
            <AudioButton
              handleClick={() => toggleCurrentSongId(id)}
              isSongPlaying={isSongPlaying && currentSongId === id}
              currentSongId={currentSongId}
            />
          </AudioButtonWrapper>
          <ButtonWrapper>
            <Button
              id={id}
              handleClick={() => saveSong(id)}
              disabled={savedSongsList?.some((song) => song.id === id)}
            >
              Save this song
            </Button>
          </ButtonWrapper>
        </ListItemStyled>
      ))}
    </ListStyled>
  )
}

const ListStyled = styled.ul`
  display: grid;
  gap: 8px;
  list-style: none;
  margin: 0;
  padding: 0;
`
const ListItemStyled = styled.li`
  background-color: var(--primary-light);
  display: grid;
  grid-template-columns: 80% auto;
  grid-template-rows: 1fr 1fr 2fr;
  padding: 20px;
`
const Artist = styled.span`
  display: block;
  font-size: 75%;
  font-weight: 300;
  padding-bottom: 6px;
  text-transform: uppercase;
`
const Song = styled.span`
  color: var(--primary-dark);
  font-size: 1em;
  font-weight: 600;
  grid-column-start: 1;
`
const AudioButtonWrapper = styled.div`
  grid-column-start: 2;
  grid-row: 1/3;
  justify-self: end;
`
const ButtonWrapper = styled.div`
  align-self: end;
  grid-column: 1/3;
`
