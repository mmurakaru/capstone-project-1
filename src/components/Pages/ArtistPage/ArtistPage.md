One list item:

```js
<ArtistList
  artists={[
    {
      artist: 'Unprocessed',
      songTitle: 'Real',
      songSnippet: 'https://test/',
      id: 'a',
    },
  ]}
  toggleCurrentSongId={() => {}}
  isSongPlaying="false"
  currentSongId="a"
  saveSong={() => {}}
  savedSongs={[]}
/>
```
