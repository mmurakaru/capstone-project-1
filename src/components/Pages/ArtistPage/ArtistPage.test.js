import ArtistPage from './ArtistPage'
import { render } from '@testing-library/react'
import user from '@testing-library/user-event'

describe('ArtistPage', () => {
  beforeEach(() => {
    window.HTMLMediaElement.prototype.play = jest.fn()
    window.HTMLMediaElement.prototype.pause = jest.fn()
    window.HTMLMediaElement.prototype.load = jest.fn()
  })
  it('renders correctly', () => {
    const { container } = render(
      <ArtistPage
        artists={[
          {
            artist: 'Unprocessed',
            songTitle: 'Real',
            songSnippet: 'https://test/',
            id: 'a',
          },
        ]}
        toggleCurrentSongId={() => {}}
        isSongPlaying={false}
        currentSongId="a"
        saveSong={() => {}}
        savedSongs={[]}
      />
    )
    expect(container.firstChild).toMatchSnapshot()
  })
  it('sets the current id to currentSongId', () => {
    const toggleCurrentSongIdMock = jest.fn()
    const { getByTestId } = render(
      <ArtistPage
        artists={[
          {
            artist: 'Unprocessed',
            songTitle: 'Real',
            songSnippet: 'https://test/',
            id: 'a',
          },
        ]}
        toggleCurrentSongId={toggleCurrentSongIdMock}
        isSongPlaying={false}
        currentSongId="a"
        saveSong={() => {}}
        savedSongs={[]}
      />
    )
    const button = getByTestId('audio-button')
    user.click(button)
    expect(toggleCurrentSongIdMock).toHaveBeenCalled()
  })
  it('toggles the isSongPlaying state correctly', () => {
    const { getByTitle, queryByTitle, rerender } = render(
      <ArtistPage
        artists={[
          {
            artist: 'Unprocessed',
            songTitle: 'Real',
            songSnippet: 'https://test/',
            id: 'a',
          },
        ]}
        toggleCurrentSongId={() => {}}
        isSongPlaying={false}
        currentSongId="a"
        saveSong={() => {}}
        savedSongs={[]}
      />
    )
    expect(getByTitle('play')).toBeInTheDocument()

    rerender(
      <ArtistPage
        artists={[
          {
            artist: 'Unprocessed',
            songTitle: 'Real',
            songSnippet: 'https://test/',
            id: 'a',
          },
        ]}
        toggleCurrentSongId={() => {}}
        isSongPlaying={true}
        currentSongId="a"
        saveSong={() => {}}
        savedSongs={[]}
      />
    )
    expect(queryByTitle('play')).not.toBeInTheDocument()
    expect(getByTitle('pause')).toBeInTheDocument()
  })
  it('calls the saveSong function correctly', () => {
    const saveSongMock = jest.fn()
    const { getByTestId } = render(
      <ArtistPage
        artists={[
          {
            artist: 'Unprocessed',
            songTitle: 'Real',
            songSnippet: 'https://test/',
            id: 'a',
          },
        ]}
        toggleCurrentSongId={() => {}}
        isSongPlaying={false}
        currentSongId="a"
        saveSong={saveSongMock}
        savedSongs={[]}
      />
    )
    const button = getByTestId('button')
    user.click(button)
    expect(saveSongMock).toHaveBeenCalled()
  })
  it('renders a new array for savedSongs by calling saveSong function', () => {
    const saveSongMock = jest.fn()
    const { getByTestId, rerender } = render(
      <ArtistPage
        artists={[
          {
            artist: 'Unprocessed',
            songTitle: 'Real',
            songSnippet: 'https://test/',
            id: 'a',
          },
        ]}
        toggleCurrentSongId={() => {}}
        isSongPlaying={false}
        currentSongId="a"
        saveSong={saveSongMock}
        savedSongs={[]}
      />
    )
    const button = getByTestId('button')
    user.click(button)
    expect(saveSongMock).toHaveBeenCalled()

    rerender(
      <ArtistPage
        artists={[
          {
            artist: 'Unprocessed',
            songTitle: 'Real',
            songSnippet: 'https://test/',
            id: 'a',
          },
        ]}
        toggleCurrentSongId={() => {}}
        isSongPlaying={false}
        currentSongId="a"
        saveSong={saveSongMock}
        savedSongs={[
          {
            artist: 'Unprocessed',
            songTitle: 'Real',
            songSnippet: 'https://test/',
            id: 'a',
          },
        ]}
      />
    )
  })
})
