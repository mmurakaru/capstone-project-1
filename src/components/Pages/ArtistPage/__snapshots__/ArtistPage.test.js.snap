// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtistPage renders correctly 1`] = `
.c5 {
  background-color: transparent;
  border: var(--primary-dark) 2px solid;
  fill: var(--primary-dark);
  height: 46px;
  padding: 10px;
  width: 46px;
}

.c7 {
  background-color: transparent;
  border: var(--primary-dark) 2px solid;
  color: var(--primary-dark);
  font-size: 0.8em;
  padding: 6px;
  width: 100%;
}

.c7:disabled {
  opacity: 50%;
}

.c0 {
  display: grid;
  gap: 8px;
  list-style: none;
  margin: 0;
  padding: 0;
}

.c1 {
  background-color: var(--primary-light);
  display: grid;
  grid-template-columns: 80% auto;
  grid-template-rows: 1fr 1fr 2fr;
  padding: 20px;
}

.c2 {
  display: block;
  font-size: 75%;
  font-weight: 300;
  padding-bottom: 6px;
  text-transform: uppercase;
}

.c3 {
  color: var(--primary-dark);
  font-size: 1em;
  font-weight: 600;
  grid-column-start: 1;
}

.c4 {
  grid-column-start: 2;
  grid-row: 1/3;
  justify-self: end;
}

.c6 {
  -webkit-align-self: end;
  -ms-flex-item-align: end;
  align-self: end;
  grid-column: 1/3;
}

<ul
  class="c0"
>
  <li
    class="c1"
  >
    <span
      class="c2"
    >
      Unprocessed
    </span>
    <span
      class="c3"
    >
      Real
    </span>
    <div
      class="c4"
    >
      <button
        class="c5"
        data-testid="audio-button"
      >
        <svg
          title="play"
        >
          play.svg
        </svg>
      </button>
    </div>
    <div
      class="c6"
    >
      <button
        class="c7"
        data-testid="button"
      >
        Save this song
      </button>
    </div>
  </li>
</ul>
`;
